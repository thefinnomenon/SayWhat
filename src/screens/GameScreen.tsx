import React, { useState } from 'react';
import { Theme } from '../../types';
import styled from 'styled-components/native';
import { Button } from '../components/Button';
import { useNavigationParam } from 'react-navigation-hooks';
import { Button as Btn, Alert } from 'react-native';

export const GameScreen = () => {
  const words = useNavigationParam('words');

  const getRandomWord = () => {
    return words[Math.round(Math.round(Math.random() * (words.length - 1)))];
  };

  const [word, setWord] = useState(getRandomWord());

  return (
    <>
      <Container>
        <WordContainer>
          <Word>{word}</Word>
        </WordContainer>
        <Spacer flex={3} />
      </Container>
      <BottomButton title="NEXT" onPress={() => setWord(getRandomWord())} />
    </>
  );
};

// @ts-ignore: react-navigation has messed up types
const quitConfirmation = navigation =>
  Alert.alert(
    'Quit Confirmation',
    'Are you sure you want to quit the current game?',
    [
      {
        text: 'Cancel',
        style: 'cancel',
      },
      {
        text: 'Yes',
        onPress: () => navigation.navigate('Home'),
      },
    ],
    { cancelable: true },
  );

const Container = styled.View<Theme>`
  flex: 1;
  justify-content: center;
  align-items: center;
  background: ${props => props.theme.colors.background};
`;

const WordContainer = styled.View`
  flex: 6;
  justify-content: center;
  align-items: center;
`;

const Word = styled.Text<Theme>`
  color: ${props => props.theme.colors.text};
  font-size: 48;
  font-weight: bold;
  margin: 8px;
`;

const BottomButton = styled(Button)<Theme>`
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  height: 80px;
  background: ${props => props.theme.colors.primary};
`;

type SpacerProps = {
  flex: number;
};
const Spacer = styled.View<SpacerProps>`
  flex: ${props => props.flex};
`;

// @ts-ignore: react-navigation has messed up types
GameScreen.navigationOptions = ({ navigation, screenProps }) => ({
  title: navigation.getParam('title', ''),
  headerTintColor: screenProps.theme.colors.text,
  headerStyle: {
    backgroundColor: screenProps.theme.colors.background,
    borderBottomWidth: 0,
  },
  headerLeft: (
    <Btn
      onPress={() => quitConfirmation(navigation)}
      title="Quit"
      color={screenProps.theme.colors.textAlt}
    />
  ),
});

export default GameScreen;
