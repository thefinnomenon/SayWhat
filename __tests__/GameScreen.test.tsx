import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import ThemeManager from '../contexts/ManageThemeContext';
import { I18nextProvider } from 'react-i18next';
import i18 from '../i18n';

import GameScreen from '../src/screens/GameScreen';

jest.mock('react-navigation-hooks', () => {
  return {
    useNavigationParam: jest.fn().mockImplementation(param => {
      switch (param) {
        case 'words':
          return ['Word1', 'Word2', 'Word3'];
      }
    }),
  };
});

const mathRandomSpy = jest.spyOn(Math, 'random');
mathRandomSpy.mockImplementation(() => 0.5);

const renderComponent = () =>
  renderer
    .create(
      <ThemeManager>
        <I18nextProvider i18n={i18}>
          <GameScreen />
        </I18nextProvider>
      </ThemeManager>,
    )
    .toJSON();

it('renders correctly', () => {
  const tree = renderComponent();
  expect(tree).toMatchSnapshot();
});
