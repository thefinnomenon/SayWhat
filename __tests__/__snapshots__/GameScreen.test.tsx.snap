// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexGrow": 6,
            "flexShrink": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#202020",
              "fontSize": 48,
              "fontWeight": "bold",
              "marginBottom": 8,
              "marginLeft": 8,
              "marginRight": 8,
              "marginTop": 8,
            },
          ]
        }
      >
        Word2
      </Text>
    </View>
    <View
      flex={3}
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 3,
            "flexShrink": 1,
          },
        ]
      }
    />
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignContent": "center",
        "backgroundColor": "#50b7ff",
        "borderRadius": 10,
        "bottom": 0,
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "height": 80,
        "justifyContent": "center",
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontSize": 18,
            "fontWeight": "bold",
            "textAlign": "center",
          },
        ]
      }
    >
      NEXT
    </Text>
  </View>,
]
`;
